workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,221 @@
+            }
+
+            behavioursToEnable.Clear();
         }
 
         public void DisableSpatialOSBehaviours()
         {
-            // disable + deinject all spatialOSBehavioursToDisable
+            foreach (var behaviour in behavioursToDisable)
+            {
+                try
+                {
+                    behaviour.enabled = false;
+                }
+                catch (Exception e)
+                {
+                    logger.HandleLog(LogType.Error,
+                        new LogEvent("Exception thrown in OnDisable() method of MonoBehaviour.")
+                            .WithField(LoggingUtils.LoggerName, LoggerName)
+                            .WithField(LoggingUtils.EntityId, spatialId)
+                            .WithException(e));
+                }
+
+                behaviourLibrary.DeInjectAllReadersWriters(behaviour);
+                foreach (var idToReaderWriter in behaviourToReadersWriters[behaviour])
+                {
+                    var id = idToReaderWriter.Key;
+                    var reader = idToReaderWriter.Value;
+                    compIdToReadersWriters[id].Remove(reader);
+                }
+
+                behaviourToReadersWriters.Remove(behaviour);
+            }
+
+            behavioursToDisable.Clear();
         }
 
         public void AddComponent(uint componentId)
         {
-            // Mark readers as to be enabled
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentsReady(relevantReaderSpatialOSBehaviours);
         }
 
         public void RemoveComponent(uint componentId)
         {
-            // Mark readers as to be disabled
+            // Mark reader components not ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentsNotReady(relevantReaderSpatialOSBehaviours);
         }
 
         public void ChangeAuthority(uint componentId, Authority authority)
         {
-            // Mark writers as to be enabled/disabled
+            if (authority == Authority.Authoritative)
+            {
+                // Mark writer components ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = behavioursRequiringWriterTypes[componentId];
+                MarkComponentsReady(relevantWriterSpatialOSBehaviours);
+            }
+            else if (authority == Authority.NotAuthoritative)
+            {
+                // Mark writer components not ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = behavioursRequiringWriterTypes[componentId];
+                MarkComponentsNotReady(relevantWriterSpatialOSBehaviours);
+            }
+        }
+
+        private void MarkComponentsReady(IEnumerable<MonoBehaviour> behaviours)
```

Comment:
The method name might be misleading as people might confuse "marking a requirement as satisfied" with "marking a monobehaviour to be enabled as all requirements are satisfied". How about renaming this method to `SatisfyReaderWriterRequirement()`?